File: resources/ios/IconButtonRenderer.swift
import Combine
import SwiftUI

enum IconButtonRendererEvent: Equatable {
    case press(callbackID: Int, nodeID: Int)
}

struct IconButtonRendererConfiguration: Equatable {
    let nodeID: Int
    let icon: String
    let iconVariant: String
    let variant: FirstlightIconButtonVariant
    let size: FirstlightIconButtonSize
    let disabled: Bool
    let loading: Bool
    let accessibilityLabel: String
    let accessibilityHint: String
    let callbackID: Int

    init(node: NativeUINode) {
        nodeID = node.id
        icon = node.props.getString("icon")
        iconVariant = node.props.getString("icon_variant")
        variant = FirstlightIconButtonVariant(
            rawValue: node.props.getString("variant", default: "primary")
        ) ?? .primary
        size = FirstlightIconButtonSize(
            rawValue: node.props.getString("size", default: "md")
        ) ?? .medium
        disabled = node.props.getBool("disabled")
        loading = node.props.getBool("loading")
        accessibilityLabel = node.props.getString("a11y_label")
        accessibilityHint = node.props.getString("a11y_hint")
        callbackID = node.onPress
    }

    var isEnabled: Bool {
        !disabled && !loading && callbackID != 0
    }

    func pressEvent() -> IconButtonRendererEvent? {
        guard isEnabled else { return nil }

        return .press(callbackID: callbackID, nodeID: nodeID)
    }
}

struct IconButtonRendererState {
    private(set) var configuration: IconButtonRendererConfiguration

    init(node: NativeUINode) {
        configuration = IconButtonRendererConfiguration(node: node)
    }

    @discardableResult
    mutating func serverPublished(_ tree: NativeUITree) -> Bool {
        guard let node = iconButtonNode(id: configuration.nodeID, in: tree.root) else {
            return false
        }

        let published = IconButtonRendererConfiguration(node: node)
        guard published != configuration else { return false }

        configuration = published
        return true
    }
}

struct IconButtonRenderer: View {
    let node: NativeUINode

    @ObservedObject private var themeStore = NativeUITheme.shared
    @Environment(\.colorScheme) private var colorScheme
    @State private var state: IconButtonRendererState

    init(node: NativeUINode) {
        self.node = node
        _state = State(initialValue: IconButtonRendererState(node: node))
    }

    var body: some View {
        let configuration = state.configuration

        FirstlightIconButtonControl(
            configuration: configuration,
            tokens: themeStore.resolve(for: colorScheme),
            onPress: {
                guard case let .press(callbackID, nodeID) = configuration.pressEvent() else {
                    return
                }

                NativeUIBridge.sendPressEvent(callbackID, nodeId: nodeID)
            }
        )
        .onReceive(NativeUIBridge.shared.$currentTree.compactMap { $0 }) { tree in
            state.serverPublished(tree)
        }
    }
}

private func iconButtonNode(id: Int, in node: NativeUINode) -> NativeUINode? {
    if node.id == id { return node }

    for child in node.children {
        if let found = iconButtonNode(id: id, in: child) {
            return found
        }
    }

    return nil
}

File: resources/android/IconButtonRenderer.kt
package dev.firstlightui.plugins.firstlight_ui.ui

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import com.nativephp.mobile.ui.nativerender.NativeUIBridge
import com.nativephp.mobile.ui.nativerender.NativeUINode
import com.nativephp.mobile.ui.nativerender.NativeUITree
import com.nativephp.plugins.native_ui.NativeUITheme

sealed interface IconButtonRendererEvent {
    data class Press(val callbackId: Int, val nodeId: Int) : IconButtonRendererEvent
}

data class IconButtonRendererConfiguration(
    val nodeId: Int,
    val icon: String,
    val iconVariant: String,
    val variant: FirstlightIconButtonVariant,
    val size: FirstlightIconButtonSize,
    val disabled: Boolean,
    val loading: Boolean,
    val accessibilityLabel: String,
    val accessibilityHint: String,
    val callbackId: Int,
) {
    constructor(node: NativeUINode) : this(
        nodeId = node.id,
        icon = node.props.getString("icon"),
        iconVariant = node.props.getString("icon_variant"),
        variant = FirstlightIconButtonVariant.fromWire(
            node.props.getString("variant", "primary"),
        ),
        size = FirstlightIconButtonSize.fromWire(node.props.getString("size", "md")),
        disabled = node.props.getBool("disabled"),
        loading = node.props.getBool("loading"),
        accessibilityLabel = node.props.getString("a11y_label"),
        accessibilityHint = node.props.getString("a11y_hint"),
        callbackId = node.onPress,
    )

    val enabled: Boolean
        get() = !disabled && !loading && callbackId != 0

    fun pressEvent(): IconButtonRendererEvent.Press? = if (enabled) {
        IconButtonRendererEvent.Press(callbackId = callbackId, nodeId = nodeId)
    } else {
        null
    }
}

class IconButtonRendererState(node: NativeUINode) {
    var configuration by mutableStateOf(IconButtonRendererConfiguration(node))
        private set

    fun serverPublished(tree: NativeUITree): Boolean {
        val node = tree.root.findIconButtonNode(configuration.nodeId) ?: return false
        val published = IconButtonRendererConfiguration(node)
        if (published == configuration) return false

        configuration = published
        return true
    }
}

object IconButtonRenderer {
    @Composable
    fun Render(node: NativeUINode, modifier: Modifier) {
        val state = remember(node.id) { IconButtonRendererState(node) }
        val publishedTree = NativeUIBridge.currentTree.value
        LaunchedEffect(publishedTree) {
            publishedTree?.let(state::serverPublished)
        }

        val tokens = if (isSystemInDarkTheme()) NativeUITheme.dark else NativeUITheme.light
        val configuration = state.configuration

        FirstlightIconButtonControl(
            configuration = configuration,
            tokens = tokens,
            onPress = {
                configuration.pressEvent()?.let { event ->
                    NativeUIBridge.sendPressEvent(event.callbackId, event.nodeId)
                }
            },
            modifier = modifier,
        )
    }
}

private fun NativeUINode.findIconButtonNode(id: Int): NativeUINode? {
    if (this.id == id) return this
    children.forEach { child -> child.findIconButtonNode(id)?.let { return it } }
    return null
}
